src/ScheduleFunctions.cpp

```diff
@@ -1138,149 +949,210 @@ class ShiftLoopNest : public IRMutator2 {
     }
 
 public:
-    ShiftLoopNest(const map<string, Expr> &s) : shifts(s) {}
+    explicit ShiftLoopNest(const map<string, Expr> &s) : shifts(s) {}
+
+    template<typename T>
+    static T apply_shift(const map<string, Expr> &shifts, const T &node) {
+        if (shifts.empty()) {
+            return node;
+        }
+        ShiftLoopNest visitor(shifts);
+        return visitor.mutate(node);
+    }
 };
 
-// Inject the allocation and realization of a group of functions which are
-// to be fused into an existing loop nest using its schedule.
-class InjectGroupRealization : public IRMutator2 {
+class InjectFunctionRealization : public IRMutator2 {
 public:
-    // Member of the fused loop starting from the first to be realized to the last.
-    const vector<Function> &group;
-    // List of booleans indicating if group[i] is an output
-    const vector<bool> &is_output_list;
-    bool found_store_level, found_compute_level;
-    const Target &target;
-    const map<string, Function> &env;
-    LoopLevel compute_level;
-    LoopLevel store_level;
+    InjectFunctionRealization(const vector<Function> &funcs, const vector<bool> &is_output_list, const Target &target,
+                              const map<string, Function> &env)
+            : funcs(funcs)
+            , is_output_list(is_output_list)
+            , target(target)
+            , env(env)
+            , compute_level(funcs[0].schedule().compute_level())
+            , store_level(funcs[0].schedule().store_level()) { }
+
+    bool found_compute_level() const { return _found_compute_level; }
+    bool found_store_level() const { return _found_store_level; }
+
+protected:
+    bool _found_compute_level{};
+    bool _found_store_level{};
 
-    InjectGroupRealization(const vector<Function> &g, const vector<bool> &o,
-                           const Target &t, const map<string, Function> &env)
-            : group(g), is_output_list(o), found_store_level(false),
-              found_compute_level(false), target(t), env(env) {
-        internal_assert(!group.empty());
-        internal_assert(group.size() == is_output_list.size());
+    using IRMutator2::visit;
 
-        compute_level = group[0].schedule().compute_level();
-        store_level = group[0].schedule().store_level();
-        internal_assert(!compute_level.is_inlined());
-    }
+    Stmt visit(const For *for_loop) override {
+        debug(3) << "Injecting " << funcs << " entering for-loop over " << for_loop->name << "\n";
+        Stmt body = for_loop->body;
 
-private:
+        // Dig through any placeholder prefetches
+        vector<tuple<string, vector<Type>, PrefetchDirective>> placeholder_prefetches;
```

Comment:
Make a local struct with meaningful names instead of using a std::tuple

Reply:
I'll make that improvement, but, again, not my code.